File: recipe_catalog.go
// recipe_catalog.go — Go версия

package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Recipe struct {
	ID           int      `json:"id"`
	Name         string   `json:"name"`
	Category     string   `json:"category"`
	Ingredients  []string `json:"ingredients"`
	Instructions string   `json:"instructions"`
	PrepTime     int      `json:"prep_time"`
	Difficulty   string   `json:"difficulty"`
	Rating       float64  `json:"rating"`
	RatingsCount int      `json:"ratings_count"`
}

type Catalog struct {
	Recipes []Recipe `json:"recipes"`
	file    string
}

func NewCatalog(file string) *Catalog {
	c := &Catalog{file: file}
	c.load()
	return c
}

func (c *Catalog) load() {
	data, err := os.ReadFile(c.file)
	if err != nil {
		c.Recipes = []Recipe{}
		return
	}
	json.Unmarshal(data, &c.Recipes)
}

func (c *Catalog) save() {
	data, _ := json.MarshalIndent(c.Recipes, "", "  ")
	os.WriteFile(c.file, data, 0644)
}

func (c *Catalog) add(name, category string, ingredients []string, instructions string, prepTime int, difficulty string) int {
	id := len(c.Recipes) + 1
	c.Recipes = append(c.Recipes, Recipe{
		ID:           id,
		Name:         name,
		Category:     category,
		Ingredients:  ingredients,
		Instructions: instructions,
		PrepTime:     prepTime,
		Difficulty:   difficulty,
		Rating:       0,
		RatingsCount: 0,
	})
	c.save()
	return id
}

func (c *Catalog) listAll() {
	if len(c.Recipes) == 0 {
		fmt.Println("\u001B[33mКаталог пуст.\u001B[0m")
		return
	}
	fmt.Printf("\u001B[36m%-4s %-25s %-12s %-8s %-10s %-8s\u001B[0m\n", "ID", "Название", "Категория", "Время", "Сложность", "Рейтинг")
	fmt.Println(strings.Repeat("-", 75))
	for _, r := range c.Recipes {
		ratingStr := "—"
		if r.RatingsCount > 0 {
			ratingStr = fmt.Sprintf("%.1f★", r.Rating)
		}
		diffColor := "\u001B[32m"
		if r.Difficulty == "средняя" {
			diffColor = "\u001B[33m"
		} else if r.Difficulty == "сложная" {
			diffColor = "\u001B[31m"
		}
		fmt.Printf("%-4d %-25s %-12s %-8d %s%-10s\u001B[0m %-8s\n", r.ID, r.Name, r.Category, r.PrepTime, diffColor, r.Difficulty, ratingStr)
	}
}

func (c *Catalog) search(query string) {
	query = strings.ToLower(query)
	results := []Recipe{}
	for _, r := range c.Recipes {
		if strings.Contains(strings.ToLower(r.Name), query) {
			results = append(results, r)
		} else {
			for _, ing := range r.Ingredients {
				if strings.Contains(strings.ToLower(ing), query) {
					results = append(results, r)
					break
				}
			}
		}
	}
	if len(results) == 0 {
		fmt.Println("\u001B[33mНичего не найдено.\u001B[0m")
		return
	}
	for _, r := range results {
		fmt.Printf("%d: %s | %s | %d мин | %s | %.1f★\n", r.ID, r.Name, r.Category, r.PrepTime, r.Difficulty, r.Rating)
	}
}

func (c *Catalog) filterBy(field, value string) {
	results := []Recipe{}
	for _, r := range c.Recipes {
		switch field {
		case "category":
			if strings.EqualFold(r.Category, value) {
				results = append(results, r)
			}
		case "difficulty":
			if strings.EqualFold(r.Difficulty, value) {
				results = append(results, r)
			}
		case "prep_time":
			if v, err := strconv.Atoi(value); err == nil && r.PrepTime <= v {
				results = append(results, r)
			}
		}
	}
	if len(results) == 0 {
		fmt.Println("\u001B[33mНет рецептов, соответствующих фильтру.\u001B[0m")
		return
	}
	for _, r := range results {
		fmt.Printf("%d: %s | %s | %d мин | %s\n", r.ID, r.Name, r.Category, r.PrepTime, r.Difficulty)
	}
}

func (c *Catalog) sortBy(field string, reverse bool) {
	switch field {
	case "name":
		if reverse {
			SortBy(c.Recipes, func(i, j int) bool { return c.Recipes[i].Name > c.Recipes[j].Name })
		} else {
			SortBy(c.Recipes, func(i, j int) bool { return c.Recipes[i].Name < c.Recipes[j].Name })
		}
	case "prep_time":
		if reverse {
			SortBy(c.Recipes, func(i, j int) bool { return c.Recipes[i].PrepTime > c.Recipes[j].PrepTime })
		} else {
			SortBy(c.Recipes, func(i, j int) bool { return c.Recipes[i].PrepTime < c.Recipes[j].PrepTime })
		}
	case "rating":
		if reverse {
			SortBy(c.Recipes, func(i, j int) bool { return c.Recipes[i].Rating > c.Recipes[j].Rating })
		} else {
			SortBy(c.Recipes, func(i, j int) bool { return c.Recipes[i].Rating < c.Recipes[j].Rating })
		}
	default:
		fmt.Println("\u001B[31mНеверное поле для сортировки.\u001B[0m")
		return
	}
	c.listAll()
}

func SortBy(recipes []Recipe, less func(i, j int) bool) {
	for i := 0; i < len(recipes)-1; i++ {
		for j := i + 1; j < len(recipes); j++ {
			if less(i, j) {
				recipes[i], recipes[j] = recipes[j], recipes[i]
			}
		}
	}
}

func (c *Catalog) delete(id int) bool {
	for i, r := range c.Recipes {
		if r.ID == id {
			c.Recipes = append(c.Recipes[:i], c.Recipes[i+1:]...)
			c.save()
			return true
		}
	}
	return false
}

func (c *Catalog) edit(id int, field, value string) bool {
	for i, r := range c.Recipes {
		if r.ID == id {
			switch field {
			case "name":
				c.Recipes[i].Name = value
			case "category":
				c.Recipes[i].Category = value
			case "ingredients":
				c.Recipes[i].Ingredients = strings.Split(value, ",")
				for j := range c.Recipes[i].Ingredients {
					c.Recipes[i].Ingredients[j] = strings.TrimSpace(c.Recipes[i].Ingredients[j])
				}
			case "instructions":
				c.Recipes[i].Instructions = value
			case "prep_time":
				if v, err := strconv.Atoi(value); err == nil {
					c.Recipes[i].PrepTime = v
				} else {
					return false
				}
			case "difficulty":
				c.Recipes[i].Difficulty = value
			default:
				return false
			}
			c.save()
			return true
		}
	}
	return false
}

func (c *Catalog) rate(id int, rating int) bool {
	for i, r := range c.Recipes {
		if r.ID == id {
			total := r.Rating*float64(r.RatingsCount) + float64(rating)
			c.Recipes[i].RatingsCount++
			c.Recipes[i].Rating = total / float64(c.Recipes[i].RatingsCount)
			c.save()
			return true
		}
	}
	return false
}

func (c *Catalog) stats() {
	if len(c.Recipes) == 0 {
		fmt.Println("Нет данных.")
		return
	}
	total := len(c.Recipes)
	categories := make(map[string]int)
	difficulties := make(map[string]int)
	totalRating := 0.0
	ratedCount := 0
	for _, r := range c.Recipes {
		categories[r.Category]++
		difficulties[r.Difficulty]++
		if r.RatingsCount > 0 {
			totalRating += r.Rating
			ratedCount++
		}
	}
	avgRating := 0.0
	if ratedCount > 0 {
		avgRating = totalRating / float64(ratedCount)
	}
	fmt.Println("\u001B[36m📊 Статистика:\u001B[0m")
	fmt.Printf("  Всего рецептов: %d\n", total)
	fmt.Printf("  Средний рейтинг: %.1f★ (из %d оценённых)\n", avgRating, ratedCount)
	fmt.Println("  По категориям:")
	for c, cnt := range categories {
		fmt.Printf("    %s: %d\n", c, cnt)
	}
	fmt.Println("  По сложности:")
	for d, cnt := range difficulties {
		fmt.Printf("    %s: %d\n", d, cnt)
	}
}

func (c *Catalog) exportJSON(filename string) {
	data, _ := json.MarshalIndent(c.Recipes, "", "  ")
	os.WriteFile(filename, data, 0644)
	fmt.Printf("\u001B[32m💾 Экспорт JSON: %s\u001B[0m\n", filename)
}

func (c *Catalog) exportCSV(filename string) {
	if len(c.Recipes) == 0 {
		fmt.Println("\u001B[33mНет данных для экспорта.\u001B[0m")
		return
	}
	file, _ := os.Create(filename)
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{"ID", "Название", "Категория", "Ингредиенты", "Инструкция", "Время", "Сложность", "Рейтинг"})
	for _, r := range c.Recipes {
		writer.Write([]string{
			strconv.Itoa(r.ID),
			r.Name,
			r.Category,
			strings.Join(r.Ingredients, ", "),
			r.Instructions,
			strconv.Itoa(r.PrepTime),
			r.Difficulty,
			fmt.Sprintf("%.1f", r.Rating),
		})
	}
	fmt.Printf("\u001B[32m💾 Экспорт CSV: %s\u001B[0m\n", filename)
}

func main() {
	catalog := NewCatalog("recipes.json")
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Println("\n\u001B[36m🍰 Dessert Recipe Catalog (Go)\u001B[0m")
		fmt.Println("1. Добавить рецепт")
		fmt.Println("2. Показать все рецепты")
		fmt.Println("3. Поиск рецептов")
		fmt.Println("4. Фильтрация")
		fmt.Println("5. Сортировка")
		fmt.Println("6. Удалить рецепт")
		fmt.Println("7. Редактировать рецепт")
		fmt.Println("8. Оценить рецепт")
		fmt.Println("9. Статистика")
		fmt.Println("10. Экспорт")
		fmt.Println("11. Выход")
		fmt.Print("Выберите действие: ")
		choice, _ := reader.ReadString('\n')
		choice = strings.TrimSpace(choice)
		switch choice {
		case "1":
			fmt.Print("Название: ")
			name, _ := reader.ReadString('\n')
			name = strings.TrimSpace(name)
			fmt.Print("Категория (торт, пирожное, печенье, мороженое, другое): ")
			category, _ := reader.ReadString('\n')
			category = strings.TrimSpace(category)
			fmt.Print("Ингредиенты (через запятую): ")
			ingStr, _ := reader.ReadString('\n')
			ingredients := strings.Split(ingStr, ",")
			for j := range ingredients {
				ingredients[j] = strings.TrimSpace(ingredients[j])
			}
			fmt.Print("Инструкция: ")
			instructions, _ := reader.ReadString('\n')
			instructions = strings.TrimSpace(instructions)
			fmt.Print("Время приготовления (мин): ")
			timeStr, _ := reader.ReadString('\n')
			prepTime, _ := strconv.Atoi(strings.TrimSpace(timeStr))
			fmt.Print("Сложность (лёгкая/средняя/сложная): ")
			difficulty, _ := reader.ReadString('\n')
			difficulty = strings.TrimSpace(strings.ToLower(difficulty))
			if difficulty != "лёгкая" && difficulty != "средняя" && difficulty != "сложная" {
				difficulty = "средняя"
			}
			id := catalog.add(name, category, ingredients, instructions, prepTime, difficulty)
			fmt.Printf("\u001B[32m✅ Рецепт добавлен (ID: %d)\u001B[0m\n", id)
		case "2":
			catalog.listAll()
		case "3":
			fmt.Print("Введите запрос (название или ингредиент): ")
			query, _ := reader.ReadString('\n')
			query = strings.TrimSpace(query)
			catalog.search(query)
		case "4":
			fmt.Println("Фильтровать по: category, difficulty, prep_time")
			fmt.Print("Поле: ")
			field, _ := reader.ReadString('\n')
			field = strings.TrimSpace(strings.ToLower(field))
			fmt.Print("Значение: ")
			value, _ := reader.ReadString('\n')
			value = strings.TrimSpace(value)
			catalog.filterBy(field, value)
		case "5":
			fmt.Println("Сортировать по: name, prep_time, rating")
			fmt.Print("Поле: ")
			field, _ := reader.ReadString('\n')
			field = strings.TrimSpace(strings.ToLower(field))
			fmt.Print("По убыванию? (y/n): ")
			revStr, _ := reader.ReadString('\n')
			reverse := strings.TrimSpace(strings.ToLower(revStr)) == "y"
			catalog.sortBy(field, reverse)
		case "6":
			catalog.listAll()
			fmt.Print("Введите ID для удаления: ")
			idStr, _ := reader.ReadString('\n')
			id, _ := strconv.Atoi(strings.TrimSpace(idStr))
			if catalog.delete(id) {
				fmt.Println("\u001B[32m✅ Рецепт удалён.\u001B[0m")
			} else {
				fmt.Println("\u001B[31m❌ Рецепт не найден.\u001B[0m")
			}
		case "7":
			catalog.listAll()
			fmt.Print("Введите ID для редактирования: ")
			idStr, _ := reader.ReadString('\n')
			id, _ := strconv.Atoi(strings.TrimSpace(idStr))
			fmt.Print("Какое поле редактировать (name, category, ingredients, instructions, prep_time, difficulty): ")
			field, _ := reader.ReadString('\n')
			field = strings.TrimSpace(strings.ToLower(field))
			fmt.Print("Новое значение: ")
			value, _ := reader.ReadString('\n')
			value = strings.TrimSpace(value)
			if catalog.edit(id, field, value) {
				fmt.Println("\u001B[32m✅ Рецепт обновлён.\u001B[0m")
			} else {
				fmt.Println("\u001B[31m❌ Не удалось обновить.\u001B[0m")
			}
		case "8":
			catalog.listAll()
			fmt.Print("Введите ID для оценки: ")
			idStr, _ := reader.ReadString('\n')
			id, _ := strconv.Atoi(strings.TrimSpace(idStr))
			fmt.Print("Оценка (1-5): ")
			ratingStr, _ := reader.ReadString('\n')
			rating, _ := strconv.Atoi(strings.TrimSpace(ratingStr))
			if rating >= 1 && rating <= 5 {
				if catalog.rate(id, rating) {
					fmt.Println("\u001B[32m✅ Оценка добавлена.\u001B[0m")
				} else {
					fmt.Println("\u001B[31m❌ Рецепт не найден.\u001B[0m")
				}
			} else {
				fmt.Println("\u001B[31m❌ Оценка должна быть от 1 до 5.\u001B[0m")
			}
		case "9":
			catalog.stats()
		case "10":
			fmt.Println("1. Экспорт в JSON")
			fmt.Println("2. Экспорт в CSV")
			fmt.Print("Выберите формат: ")
			sub, _ := reader.ReadString('\n')
			sub = strings.TrimSpace(sub)
			if sub == "1" {
				catalog.exportJSON("recipes_export.json")
			} else if sub == "2" {
				catalog.exportCSV("recipes_export.csv")
			} else {
				fmt.Println("\u001B[31mНеверный выбор.\u001B[0m")
			}
		case "11":
			fmt.Println("До свидания!")
			return
		default:
			fmt.Println("\u001B[31mНеверный выбор.\u001B[0m")
		}
	}
}

File: recipe_catalog.cs
// recipe_catalog.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

class Recipe {
    public int Id { get; set; }
    public string Name { get; set; }
    public string Category { get; set; }
    public List<string> Ingredients { get; set; }
    public string Instructions { get; set; }
    public int PrepTime { get; set; }
    public string Difficulty { get; set; }
    public double Rating { get; set; }
    public int RatingsCount { get; set; }

    public void AddRating(int rating) {
        double total = Rating * RatingsCount + rating;
        RatingsCount++;
        Rating = total / RatingsCount;
    }
}

class RecipeCatalog {
    private List<Recipe> recipes = new List<Recipe>();
    private const string DataFile = "recipes.json";

    public RecipeCatalog() {
        Load();
    }

    private void Load() {
        if (File.Exists(DataFile)) {
            try {
                string json = File.ReadAllText(DataFile);
                recipes = JsonSerializer.Deserialize<List<Recipe>>(json) ?? new List<Recipe>();
            } catch {
                recipes = new List<Recipe>();
            }
        }
    }

    private void Save() {
        string json = JsonSerializer.Serialize(recipes, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(DataFile, json);
    }

    public int Add(string name, string category, List<string> ingredients, string instructions, int prepTime, string difficulty) {
        int id = recipes.Count + 1;
        recipes.Add(new Recipe {
            Id = id,
            Name = name,
            Category = category,
            Ingredients = ingredients,
            Instructions = instructions,
            PrepTime = prepTime,
            Difficulty = difficulty,
            Rating = 0,
            RatingsCount = 0
        });
        Save();
        return id;
    }

    public void ListAll() {
        if (recipes.Count == 0) {
            Console.WriteLine("\u001B[33mКаталог пуст.\u001B[0m");
            return;
        }
        Console.WriteLine($"\u001B[36m{"ID",-4} {"Название",-25} {"Категория",-12} {"Время",-8} {"Сложность",-10} {"Рейтинг",-8}\u001B[0m");
        Console.WriteLine(new string('-', 75));
        foreach (var r in recipes) {
            string ratingStr = r.RatingsCount > 0 ? $"{r.Rating:F1}★" : "—";
            string diffColor = r.Difficulty == "лёгкая" ? "\u001B[32m" : r.Difficulty == "средняя" ? "\u001B[33m" : "\u001B[31m";
            Console.WriteLine($"{r.Id,-4} {r.Name,-25} {r.Category,-12} {r.PrepTime,-8} {diffColor}{r.Difficulty,-10}\u001B[0m {ratingStr,-8}");
        }
    }

    public void Search(string query) {
        query = query.ToLower();
        var results = recipes.Where(r =>
            r.Name.ToLower().Contains(query) ||
            r.Ingredients.Any(i => i.ToLower().Contains(query))
        ).ToList();
        if (results.Count == 0) {
            Console.WriteLine("\u001B[33mНичего не найдено.\u001B[0m");
            return;
        }
        foreach (var r in results) {
            Console.WriteLine($"{r.Id}: {r.Name} | {r.Category} | {r.PrepTime} мин | {r.Difficulty} | {r.Rating:F1}★");
        }
    }

    public void FilterBy(string field, string value) {
        List<Recipe> results = new List<Recipe>();
        switch (field.ToLower()) {
            case "category":
                results = recipes.Where(r => r.Category.Equals(value, StringComparison.OrdinalIgnoreCase)).ToList();
                break;
            case "difficulty":
                results = recipes.Where(r => r.Difficulty.Equals(value, StringComparison.OrdinalIgnoreCase)).ToList();
                break;
            case "prep_time":
                if (int.TryParse(value, out int v)) {
                    results = recipes.Where(r => r.PrepTime <= v).ToList();
                }
                break;
        }
        if (results.Count == 0) {
            Console.WriteLine("\u001B[33mНет рецептов, соответствующих фильтру.\u001B[0m");
            return;
        }
        foreach (var r in results) {
            Console.WriteLine($"{r.Id}: {r.Name} | {r.Category} | {r.PrepTime} мин | {r.Difficulty}");
        }
    }

    public void SortBy(string field, bool reverse) {
        switch (field.ToLower()) {
            case "name":
                recipes = reverse ? recipes.OrderByDescending(r => r.Name).ToList() : recipes.OrderBy(r => r.Name).ToList();
                break;
            case "prep_time":
                recipes = reverse ? recipes.OrderByDescending(r => r.PrepTime).ToList() : recipes.OrderBy(r => r.PrepTime).ToList();
                break;
            case "rating":
                recipes = reverse ? recipes.OrderByDescending(r => r.Rating).ToList() : recipes.OrderBy(r => r.Rating).ToList();
                break;
            default:
                Console.WriteLine("\u001B[31mНеверное поле для сортировки.\u001B[0m");
                return;
        }
        ListAll();
    }

    public bool Delete(int id) {
        var recipe = recipes.FirstOrDefault(r => r.Id == id);
        if (recipe != null) {
            recipes.Remove(recipe);
            Save();
            return true;
        }
        return false;
    }

    public bool Edit(int id, string field, string value) {
        var recipe = recipes.FirstOrDefault(r => r.Id == id);
        if (recipe == null) return false;
        switch (field.ToLower()) {
            case "name": recipe.Name = value; break;
            case "category": recipe.Category = value; break;
            case "ingredients": recipe.Ingredients = value.Split(',').Select(s => s.Trim()).ToList(); break;
            case "instructions": recipe.Instructions = value; break;
            case "prep_time": if (int.TryParse(value, out int t)) recipe.PrepTime = t; else return false; break;
            case "difficulty": recipe.Difficulty = value; break;
            default: return false;
        }
        Save();
        return true;
    }

    public bool Rate(int id, int rating) {
        var recipe = recipes.FirstOrDefault(r => r.Id == id);
        if (recipe == null) return false;
        recipe.AddRating(rating);
        Save();
        return true;
    }

    public void Stats() {
        if (recipes.Count == 0) {
            Console.WriteLine("Нет данных.");
            return;
        }
        var categories = recipes.GroupBy(r => r.Category).ToDictionary(g => g.Key, g => g.Count());
        var difficulties = recipes.GroupBy(r => r.Difficulty).ToDictionary(g => g.Key, g => g.Count());
        double totalRating = recipes.Where(r => r.RatingsCount > 0).Sum(r => r.Rating);
        int ratedCount = recipes.Count(r => r.RatingsCount > 0);
        double avgRating = ratedCount > 0 ? totalRating / ratedCount : 0;
        Console.WriteLine("\u001B[36m📊 Статистика:\u001B[0m");
        Console.WriteLine($"  Всего рецептов: {recipes.Count}");
        Console.WriteLine($"  Средний рейтинг: {avgRating:F1}★ (из {ratedCount} оценённых)");
        Console.WriteLine("  По категориям:");
        foreach (var kv in categories.OrderByDescending(kv => kv.Value)) {
            Console.WriteLine($"    {kv.Key}: {kv.Value}");
        }
        Console.WriteLine("  По сложности:");
        foreach (var kv in difficulties) {
            Console.WriteLine($"    {kv.Key}: {kv.Value}");
        }
    }

    public void ExportJSON(string filename = "recipes_export.json") {
        string json = JsonSerializer.Serialize(recipes, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
        Console.WriteLine($"\u001B[32m💾 Экспорт JSON: {filename}\u001B[0m");
    }

    public void ExportCSV(string filename = "recipes_export.csv") {
        if (recipes.Count == 0) {
            Console.WriteLine("\u001B[33mНет данных для экспорта.\u001B[0m");
            return;
        }
        using var writer = new StreamWriter(filename);
        writer.WriteLine("ID,Название,Категория,Ингредиенты,Инструкция,Время,Сложность,Рейтинг");
        foreach (var r in recipes) {
            writer.WriteLine($"{r.Id},{r.Name},{r.Category},\"{string.Join(", ", r.Ingredients)}\",\"{r.Instructions}\",{r.PrepTime},{r.Difficulty},{r.Rating:F1}");
        }
        Console.WriteLine($"\u001B[32m💾 Экспорт CSV: {filename}\u001B[0m");
    }

    public static void Main() {
        var catalog = new RecipeCatalog();
        while (true) {
            Console.WriteLine("\n\u001B[36m🍰 Dessert Recipe Catalog (C#)\u001B[0m");
            Console.WriteLine("1. Добавить рецепт");
            Console.WriteLine("2. Показать все рецепты");
            Console.WriteLine("3. Поиск рецептов");
            Console.WriteLine("4. Фильтрация");
            Console.WriteLine("5. Сортировка");
            Console.WriteLine("6. Удалить рецепт");
            Console.WriteLine("7. Редактировать рецепт");
            Console.WriteLine("8. Оценить рецепт");
            Console.WriteLine("9. Статистика");
            Console.WriteLine("10. Экспорт");
            Console.WriteLine("11. Выход");
            Console.Write("Выберите действие: ");
            string choice = Console.ReadLine().Trim();
            switch (choice) {
                case "1":
                    Console.Write("Название: ");
                    string name = Console.ReadLine().Trim();
                    Console.Write("Категория (торт, пирожное, печенье, мороженое, другое): ");
                    string category = Console.ReadLine().Trim();
                    Console.Write("Ингредиенты (через запятую): ");
                    var ingredients = Console.ReadLine().Split(',').Select(s => s.Trim()).ToList();
                    Console.Write("Инструкция: ");
                    string instructions = Console.ReadLine().Trim();
                    Console.Write("Время приготовления (мин): ");
                    int prepTime = int.Parse(Console.ReadLine().Trim());
                    Console.Write("Сложность (лёгкая/средняя/сложная): ");
                    string difficulty = Console.ReadLine().Trim().ToLower();
                    if (difficulty != "лёгкая" && difficulty != "средняя" && difficulty != "сложная") difficulty = "средняя";
                    int id = catalog.Add(name, category, ingredients, instructions, prepTime, difficulty);
                    Console.WriteLine($"\u001B[32m✅ Рецепт добавлен (ID: {id})\u001B[0m");
                    break;
                case "2": catalog.ListAll(); break;
                case "3":
                    Console.Write("Введите запрос (название или ингредиент): ");
                    string query = Console.ReadLine().Trim();
                    catalog.Search(query);
                    break;
                case "4":
                    Console.WriteLine("Фильтровать по: category, difficulty, prep_time");
                    Console.Write("Поле: ");
                    string field = Console.ReadLine().Trim().ToLower();
                    Console.Write("Значение: ");
                    string value = Console.ReadLine().Trim();
                    catalog.FilterBy(field, value);
                    break;
                case "5":
                    Console.WriteLine("Сортировать по: name, prep_time, rating");
                    Console.Write("Поле: ");
                    string sortField = Console.ReadLine().Trim().ToLower();
                    Console.Write("По убыванию? (y/n): ");
                    bool reverse = Console.ReadLine().Trim().ToLower() == "y";
                    catalog.SortBy(sortField, reverse);
                    break;
                case "6":
                    catalog.ListAll();
                    Console.Write("Введите ID для удаления: ");
                    int delId = int.Parse(Console.ReadLine().Trim());
                    if (catalog.Delete(delId)) {
                        Console.WriteLine("\u001B[32m✅ Рецепт удалён.\u001B[0m");
                    } else {
                        Console.WriteLine("\u001B[31m❌ Рецепт не найден.\u001B[0m");
                    }
                    break;
                case "7":
                    catalog.ListAll();
                    Console.Write("Введите ID для редактирования: ");
                    int editId = int.Parse(Console.ReadLine().Trim());
                    Console.Write("Какое поле редактировать (name, category, ingredients, instructions, prep_time, difficulty): ");
                    string editField = Console.ReadLine().Trim().ToLower();
                    Console.Write("Новое значение: ");
                    string editValue = Console.ReadLine().Trim();
                    if (catalog.Edit(editId, editField, editValue)) {
                        Console.WriteLine("\u001B[32m✅ Рецепт обновлён.\u001B[0m");
                    } else {
                        Console.WriteLine("\u001B[31m❌ Не удалось обновить.\u001B[0m");
                    }
                    break;
                case "8":
                    catalog.ListAll();
                    Console.Write("Введите ID для оценки: ");
                    int rateId = int.Parse(Console.ReadLine().Trim());
                    Console.Write("Оценка (1-5): ");
                    int rating = int.Parse(Console.ReadLine().Trim());
                    if (rating >= 1 && rating <= 5) {
                        if (catalog.Rate(rateId, rating)) {
                            Console.WriteLine("\u001B[32m✅ Оценка добавлена.\u001B[0m");
                        } else {
                            Console.WriteLine("\u001B[31m❌ Рецепт не найден.\u001B[0m");
                        }
                    } else {
                        Console.WriteLine("\u001B[31m❌ Оценка должна быть от 1 до 5.\u001B[0m");
                    }
                    break;
                case "9": catalog.Stats(); break;
                case "10":
                    Console.WriteLine("1. Экспорт в JSON");
                    Console.WriteLine("2. Экспорт в CSV");
                    Console.Write("Выберите формат: ");
                    string sub = Console.ReadLine().Trim();
                    if (sub == "1") catalog.ExportJSON();
                    else if (sub == "2") catalog.ExportCSV();
                    else Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m");
                    break;
                case "11": Console.WriteLine("До свидания!"); return;
                default: Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m"); break;
            }
        }
    }
}
